SU2_CFD/include/gradients/computeGradientsSymmetry.hpp

```diff
@@ -0,0 +1,319 @@
+        Gradients_Velocity[iVar][iDim] = Gradients_iPoint[idx_vel + iVar][iDim];
+        Gradients_Velocity_Reflected[iVar][iDim] = 0.0;
+      }
+    }
+
+    /*--- Q' = L^T*Q*T ---*/
+    for (auto iDim = 0u; iDim < nDim; iDim++) {
+      for (auto jDim = 0u; jDim < nDim; jDim++) {
+        for (auto kDim = 0u; kDim < nDim; kDim++) {
+          for (auto mDim = 0u; mDim < nDim; mDim++) {
+            Gradients_Velocity_Reflected[iDim][jDim] +=
+                TensorMap[iDim][mDim] * TensorMap[jDim][kDim] * Gradients_Velocity[mDim][kDim];
+          }
+        }
+      }
+    }
+
+    /*--- we have aligned such that U is the direction of the normal
+     *    in 2D: dU/dy = dV/dx = 0
+     *    in 3D: dU/dy = dV/dx = 0
+     *           dU/dz = dW/dx = 0 ---*/
+    for (auto iDim = 1u; iDim < nDim; iDim++) {
+      Gradients_Velocity_Reflected[0][iDim] = 0.0;
+      Gradients_Velocity_Reflected[iDim][0] = 0.0;
+    }
+
+    for (auto iDim = 0u; iDim < nDim; iDim++) {
+      for (auto jDim = 0u; jDim < nDim; jDim++) {
+        Gradients_Velocity[iDim][jDim] = 0.0;
+      }
+    }
+
+    /*--- now transform back the corrected velocity gradients by taking the inverse again
+     * T = (L^-1)*T' ---*/
+    for (auto iDim = 0u; iDim < nDim; iDim++) {
+      for (auto jDim = 0u; jDim < nDim; jDim++) {
+        for (auto kDim = 0u; kDim < nDim; kDim++) {
+          for (auto mDim = 0u; mDim < nDim; mDim++) {
+            Gradients_Velocity[iDim][jDim] +=
+                TensorMap[mDim][iDim] * TensorMap[kDim][jDim] * Gradients_Velocity_Reflected[mDim][kDim];
+          }
+        }
+      }
+    }
+
+    for (auto iDim = 0u; iDim < nDim; iDim++) {
+      for (auto jDim = 0u; jDim < nDim; jDim++) {
+        // todo: 1->idx.velocity
+        Gradients_iPoint[iDim + idx_vel][jDim] = Gradients_Velocity[iDim][jDim];
+      }
+    }
+  }
+
+  /*--- Reflect the gradients for all scalars (we exclude velocity). --*/
+  for (auto iVar = varBegin; iVar < varEnd; iVar++) {
+    if ((idx_vel == -1) || ((idx_vel != -1) && (iVar == 0 || iVar > nDim))) {
+      /*--- project to symmetry aligned base ---*/
+      for (auto iDim = 0u; iDim < nDim; iDim++) {
+        gradPhi[iDim] = Gradients_iPoint[iVar][iDim];
+        gradPhiReflected[iDim] = 0.0;
+      }
+
+      for (auto jDim = 0u; jDim < nDim; jDim++) {
+        for (auto iDim = 0u; iDim < nDim; iDim++) {
+          /*--- map transpose T' * grad(phi) ---*/
+          gradPhiReflected[jDim] += TensorMap[jDim][iDim] * Gradients_iPoint[iVar][iDim];
+        }
+      }
+
+      for (auto iDim = 0u; iDim < nDim; iDim++) gradPhi[iDim] = 0.0;
+
+      /*--- gradient in direction normal to symmetry is cancelled ---*/
+      gradPhiReflected[0] = 0.0;
+
+      /*--- Now transform back ---*/
+      for (auto jDim = 0u; jDim < nDim; jDim++) {
+        for (auto iDim = 0u; iDim < nDim; iDim++) {
+          gradPhi[jDim] += TensorMap[iDim][jDim] * gradPhiReflected[iDim];
+        }
+      }
```

Comment:
For example, for scalars these operations are equivalent to:
```
g = T' * [0 0 0; 0 1 0; 0 0 1] * T * g
```
(setting the first component to 0 in matrix form).
This is equivalent to
```
g = (I - T(1,:)' * T(1,:)) * g
```
(which is the same as taking out the normal component)
The tangential and orthogonal directions cancel out (as desired since they are arbitrary).

After you review the tests we can look into these simplifications.

Reply:
OK, I get it I think, so replace like this:
```
      for (auto iDim = 0u; iDim < nDim; iDim++) {
        for (auto jDim = 0u; jDim < nDim; jDim++) {
          gradients_iPoint[iVar][iDim] -= TensorMap[0][iDim] * TensorMap[0][jDim] * Gradients_iPoint[iVar][jDim];
        }
      }
```

Reply:
implemented